### PR TITLE
F4 Boost, IO and BUS changes

### PR DESCRIPTION
**Pull request by sambas**
The big one

Drivers imported from BF
Tested building CC3D and Revo, only Revo flight tested
Not all targets fixed yet.

### CODE REVIEW

**Comment by digitalentity** [comment]
Ok, this one I'm not going to merge right away as its closely related to  #338 and supersedes #389.
We need to review and test this on all supported targets as the change is fairly big.
@martinbudden you might want to have a look on this one.

**Comment by martinbudden** [comment]
@digitalentity , I certainly have some comments to make about this. Don't have time at the moment, but will comment later on today.

**Comment by digitalentity** [comment]
@martinbudden, No proble, take your time. We are not in a rush (at least not yet) :smile:

**Comment by martinbudden** [comment]
I've looked at this and it seems to be a direct import of the drivers from betaflight. This approach is problematic for a number of reasons:
1. The PR is too large to effectively review
2. There are subtle differences between betaflight and iNav that may not have been taken into account, in particular
   1. Gyro sync and init is handled differently
   2. Sensor initialisation is handled differently
   3. There are differences in the implementation of filters
   4. Others that we may not be aware of.
3. The code takes all the betaflight drivers. I'm in particular thinking of the OSD and VTX drivers. There is an alternative implementation by Cleanflight, and iNav might prefer this. At least the decision should be made consciously, rather than by default
4. This approach is effectively a "break and fix" approach. That is import all the code, and then fix up any problems we find afterwards.

I'm currently in the process of converting to the new IO and in the process I'm merging the betaflight and iNav drivers. However, I'm doing a considered merge, rather than just an import. In particular this means:
1. I'm reviewing the drivers as I merge them. This is a sort of one-off opportunity - nobody is going to take the time to review all the drivers at any point in the future.
2. The merge is two-way, when comparing the drivers, its not always the case of taking the betaflight code, there have been instances where I have merged iNav code back into betaflight.
3. Taking a considered approach can reveal new opportunities. When I worked with @blckmn in this area in betaflight one thing that came out of the cooperation was the new build system, which makes it far easier to add new targets.
4. As I do the merge, I am also looking to remove target-dependent code.
5. At every stage the code is in a fully working state. My current work is incomplete, but there is nothing preventing a new release tomorrow. After a big merge like this, who knows how long it will be before we are confident that we can make a release.

A big merge like this is superficially attractive - it gets us to a state of including the code more quickly, but the more considered approach will get us to the state of a working release more quickly.

So, in short, I don't think we should merge this in.

**Comment by sambas** [comment]
I'm just tired of waiting something to happen and hack conflicts inside own branches. (looks like 8 months ago I did navrewrite flight with revo). But it seems no one has any interest even testing this.
Ofcourse I reviewed every single file I pushed here and made choice what to take and what not from bf. Maybe choice was wrong, atleast someone could comment it. Extra drivers for example can be left out from this as none of current targets use them. But anyhow, I'll continue hacking

**Comment by oleost** [comment]
@samba I think many of users are interested in this. 

If iNav had supported revo I would buy it instead of spracingf3.

**Comment by digitalentity** [comment]
@sambas, @oleost we'll get there, just not that fast.
